#! /bin/sh -e

# DP: libgcj: Lookup source file name and line number in separated
# DP: debug files found in /usr/lib/debug

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p0 < $0
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0


--- libjava/stacktrace.cc~	2005-12-01 01:10:57.000000000 +0100
+++ libjava/stacktrace.cc	2006-03-08 21:38:25.828527016 +0100
@@ -20,6 +20,11 @@
 #endif
 
 #include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
 
 #include <java/lang/Class.h>
 #include <java/lang/Long.h>
@@ -219,6 +224,29 @@
       finder->lookup (binaryName, (jlong) offset);
       *sourceFileName = finder->getSourceFile();
       *lineNum = finder->getLineNum();
+
+      if (*lineNum == -1 && info.dli_fname[0] == '/')
+	{
+	  const char *debugPrefix = "/usr/lib/debug";
+	  char *debugPath = (char *) malloc (strlen(debugPrefix)
+					     + strlen(info.dli_fname)
+					     + 2);
+
+	  if (debugPath)
+	    {
+	      strcpy (debugPath, debugPrefix);
+	      strcat (debugPath, info.dli_fname);
+	      //printf ("%s: 0x%x\n", debugPath, offset);
+	      if (!access (debugPath, R_OK))
+		{
+		  binaryName = JvNewStringUTF (debugPath);
+		  finder->lookup (binaryName, (jlong) offset);
+		  *sourceFileName = finder->getSourceFile();
+		  *lineNum = finder->getLineNum();
+		}
+	      free (debugPath);
+	    }
+	}
     }
 #endif
 }
